### ISSUE TITLE
Principals in aws_s3_bucket_policy should be ordered for plan

### ISSUE THREAD

**Issue by sblask** [hashicorp/terraform-provider-aws#22030]
This is pretty much the same as https://github.com/hashicorp/terraform-provider-aws/issues/18878 which has been assumed to be fixed and closed so I can't comment anymore. I am still seeing the problem I described in https://github.com/hashicorp/terraform-provider-aws/issues/18878, i.e. ordering haywire when modifying the principals. But also when I don't touch the principals as in this example:

I made a change like this (and only this) to my bucket policy:
```
                  ~ {
                      - Condition = {
                          - StringEquals = {
                              - s3:x-amz-acl = "bucket-owner-full-control"
                            }
                        } -> null
 ```

and am still seeing this kind of change:

```
                      ~ Principal = {
                          ~ AWS = [
                              - "arn:aws:iam::XXXXXXXXXXXXX:role/instance-a",
                              - "arn:aws:iam::XXXXXXXXXXXXX:role/instance-b",
                              - "arn:aws:iam::XXXXXXXXXXXXX:role/instance-c",
                              - "arn:aws:iam::XXXXXXXXXXXXX:role/user-b",
                              + "arn:aws:iam::XXXXXXXXXXXXX:role/user-d",
                                "arn:aws:iam::XXXXXXXXXXXXX:role/user-c",
                                "arn:aws:iam::XXXXXXXXXXXXX:role/user-a",
                              + "arn:aws:iam::XXXXXXXXXXXXX:role/user-b",
                              + "arn:aws:iam::XXXXXXXXXXXXX:role/instance-d",
                              + "arn:aws:iam::XXXXXXXXXXXXX:role/instance-c",
                              + "arn:aws:iam::XXXXXXXXXXXXX:role/instance-a",
                              + "arn:aws:iam::XXXXXXXXXXXXX:role/instance-b",
                            ]
                        }
```

### Terraform CLI and Terraform AWS Provider Version

```
$ terraform --version
Terraform v1.1.0
on darwin_amd64
+ provider registry.terraform.io/hashicorp/aws v3.68.0
+ provider registry.terraform.io/hashicorp/random v3.1.0
+ provider registry.terraform.io/hashicorp/vault v3.0.1
```

### Affected Resource(s)
aws_s3_bucket_policy

**Comment by justinretzolk** [comment]
Hey @sblask 👋 Thank you for taking the time to file this issue. So that we have all of the relevant information in one place, can you update the issue description with the rest of the information from the [bug template](https://github.com/hashicorp/terraform-provider-aws/blob/main/.github/ISSUE_TEMPLATE/Bug_Report.md)?

**Comment by sblask** [comment]
@justinretzolk I added some more information even though I think just re-opening the closed bug would have been better.

**Comment by celik0311** [comment]
I am seeing what I think is the same issue with the following details:

```
Terraform v1.0.2
on linux_amd64
+ provider registry.terraform.io/hashicorp/aws v4.5.0
```

We are using JSON files to define our policies. 
```
resource "aws_s3_bucket_policy" "foo" {
  bucket = aws_s3_bucket.foo.id
  policy = file("${path.module}/foo.json")
}
```

Once applied the policy shows no diffs in subsequent plans/applies. But once a change is made it triggers diffs in many places in the policy due to order, where it removes and re-adds elements, specifically principals. 

```
                  ~ {
                      ~ Principal = {
                          ~ AWS = [
                              - "arn:aws:sts::XXX:assumed-role/foo/bazz",
                                "arn:aws:sts::XXX:assumed-role/foo/bar",
                              + "arn:aws:sts::XXX:assumed-role/foo/bazz",
                            ]
                        }
                        # (3 unchanged elements hidden)
                    },
```
It's interesting because if I reorder existing principals in the policy no diff is generated (as expected since it's just ordering). However, if I remove one principal it triggers a diff in many places in the policy not just limited to the place I made a change.

**Comment by YakDriver** [comment]
Thank you so much for reporting this! We're sorry that the last fix (#21997) didn't fully resolve this. We are again making changes that should 🤞 fix this (#28855). However, since we cannot reproduce all scenarios, it's possible it is still not full solved. After `v4.50.0` is available, if you're still seeing issues, please let us know!
